feat: add blog post about esp_trace component (#729)
espressif/developer-portal: content/blog/2026/06/introducing-esp-trace-component/index.md

---
title: Introducing the esp_trace component
date: 2026-06-30
featureimagecaption: "What your firmware is really doing."

> summary: A new ESP-IDF component that splits application-level tracing into pluggable encoders and transports, so adding a new trace library no longer means patching the kernel. This post gives a high-level tour and points to the example that demonstrates the integration.

Reviewer: I suggest polishing the summary a bit:

```suggestion
summary: esp_trace is a new ESP-IDF component that splits application-level tracing into pluggable encoders and transports. Adding a new trace library no longer means patching the kernel. This post gives a high-level tour and goes over the example that demonstrates the integration.
```

Author: I combined mine and your version. Please check.

tags:
  - ESP-IDF
  - tracing
  - FreeRTOS
  - debugging
authors:
  - "erhan-kurubas"
---

## Introduction

> Tracing answers a different question than logging. Logging captures what your code emits to the log. Tracing records a continuous stream of events: task switches, interrupts, and queue activity from the kernel, plus any custom events the application emits. A host tool turns the stream into a timeline you can browse. You reach for it when something is slow or behaves unexpectedly and logs alone do not explain why.

Reviewer: How about streamlining the logic here:

```suggestion
Logging is a well-known concept. Logging captures what your code emits to the log. Tracing is a different concept. Tracing records a continuous stream of events: task switches, interrupts, and queue activity from the kernel, plus any custom events the application emits. A host tool turns the stream into a timeline you can browse. You reach for it when something is slow or behaves unexpectedly and logs alone do not explain why.
```

Author: I combined mine and your version. Please check.

ESP-IDF has a new component called `esp_trace`. It is a small core that owns the application-level trace session and delegates the rest of the work to two pluggable parts: an **encoder** that defines the trace format, and a **transport** that moves those bytes off the chip. Both parts are resolved from a registry at startup by name, so you can combine them freely with no integration changes.

We built it for two reasons. First, every new trace library needs its own mechanism to hook FreeRTOS events, and without a generic extension point integrators were forced to modify `FreeRTOSConfig.h` directly. Second, the trace format used to be tightly coupled to a specific transport, so sending the same data over a different link (USB-Serial-JTAG instead of JTAG, for example) meant rewriting the integration. `esp_trace` addresses both: a well-defined extension point that lets third-party trace tools ship as normal ESP-IDF components, with no kernel patches required.

> This post is a high-level tour. The detailed documentation already lives in the [component README](https://github.com/espressif/esp-idf/tree/master/components/esp_trace) and the example README, so I will keep this short and point you there for the details.

Reviewer: >and the example README

Provide a link to the example README?

Author: Link is added but please note example is not merged to master branch yet. It is under the translation state now.

Reviewer: In this case, let's publish this article after the example is merged.

Author: Merged internally. Waiting GH sync now. I will let you know when the link is live.

Author: Link is active now. Ready to merge.

![Example trace output in a serial monitor](img/example-output.webp)

The leading number is microseconds since the previous event. No host-side decoder or special tooling is required; open the USB endpoint in any serial monitor.

The `app_main.c` of the example uses only the generic API. It calls `esp_trace_start()`, lets the tasks run for a second, then calls `esp_trace_stop()` and `esp_trace_flush()`. The application code never references the encoder or transport by name; those bindings are resolved from configuration. That separation is the purpose of the indirection.

The example is designed to be copy-paste-ready for anyone integrating a new trace recorder into ESP-IDF. The full file-by-file walkthrough, the CMake setup, and a list of common pitfalls to watch for are in the example README. Rather than duplicating those details here, that is the canonical place to look.

> ## Where to go next

Reviewer: I think it is still good to briefly recap what was discussed in the article before `Where to go next`.

PLease consider adding the section `Conclusion`.

Author: Added.

---
title: Introducing the esp_trace component
date: 2026-06-01

> featureimagecaption: "What your firmware is really doing."

Reviewer: @erhankur 

Sorry, just noticed now. To the best of my knowledge, we don't have such a parameter. Why did you add it and does it serve the purpose?

Author: No. Not for purpose. Removed.

summary: esp_trace is a new ESP-IDF component that splits application-level tracing into pluggable encoders and transports, so adding a new trace library no longer means patching the kernel. This post gives a high-level tour and walks through the example that demonstrates the integration.
tags:
  - ESP-IDF
  - tracing
  - FreeRTOS
  - debugging
authors:
  - "erhan-kurubas"
---

## Introduction

> Tracing answers a different question than logging. Logging captures discrete messages your code chooses to emit. Tracing records a continuous stream of events: task switches, interrupts, and queue activity from the kernel, plus any custom events the application emits. Typically a host tool turns the stream into a timeline you can browse, though simpler trace formats can be read directly in a serial monitor. You reach for it when something is slow or behaves unexpectedly and logs alone do not explain why.

Reviewer: I proposed updates mainly because the narrative is not straightforward.

You start the paragraph with `Tracing answers a different question than logging.`

When I read the sentence, my questions are:

- What question logging answers?
- Does logging need to answer any question? My understanding is that logging is a process of storing the messages emitted by the code. The questions come when something goes wrong later.

Author: You are right. Updated once more. PTAL

Reviewer: Now the paragraph is well structured and the narrative is solid!

Thank you for improving it.

> Tracing answers a different question than logging. Logging captures discrete messages your code chooses to emit. Tracing records a continuous stream of events: task switches, interrupts, and queue activity from the kernel, plus any custom events the application emits. Typically a host tool turns the stream into a timeline you can browse, though simpler trace formats can be read directly in a serial monitor. You reach for it when something is slow or behaves unexpectedly and logs alone do not explain why.

Reviewer: > Typically a host tool turns the stream into ...

I recommend expanding `stream` to `stream of events` to improve readability.

Author: "Stream of events" is already used in the previous sentence, only one clause earlier, and "the stream" is being used as a definite back-reference.

To avoid repetition, how about using `that stream` to refer back to `stream of events`?

Reviewer: `That stream` works better. Let's keep it.

Just as a side note, the goal of technical content is to convey important information in a clear and unambiguous way. Repeating terminology helps avoid ambiguity, so there is nothing wrong with that.

But fiction literature or creating writing is a different story. Frequent repetitions in fiction make the narrative monotonous and boring.
